src/auth/auth.spec.ts

```diff
@@ -0,0 +1,250 @@
+    );
+  });
+  it('should throw Error("The token is valid, but...")', () => {
+    const token = jwtService.sign({
+      authorization: { userId: 1, permissions: [] },
+      signedAt: Date.now(),
+      //validUntil: Date.now() + 1000,
+    });
+    expect(() => authService.verify(token)).toThrow(
+      'The token is valid, but the payload of the token is' +
+        ' not a TokenPayload object. This is ether a bug or a malicious attack.',
+    );
+  });
+  it('should throw Error("The token is valid, but...")', () => {
+    const token = jwtService.sign({
+      authorization: {
+        userId: 1,
+        permissions: [
+          {
+            authorizedActions: [AuthorizedAction.create],
+            authorizedResource: {
+              resourceType: 'user',
+              resourceId: 1,
+            },
+          },
+        ],
+      },
+      signedAt: Date.now(),
+      validUntil: Date.now() + 1000,
+    });
+    expect(() => authService.verify(token)).toThrow(
+      'The token is valid, but the payload of the token is' +
+        ' not a TokenPayload object. This is ether a bug or a malicious attack.',
+    );
+  });
+  it('should throw Error("The token is valid, but...")', () => {
+    const token = jwtService.sign({
+      authorization: {
+        userId: 1,
+        permissions: [
+          {
+            authorizedActions: [AuthorizedAction.create],
+            authorizedResource: {
+              resourceType: 'user',
+              resourceId: 1,
+              ownedByUser: '1',
+            },
+          },
+        ],
+      },
+      signedAt: Date.now(),
+      validUntil: Date.now() + 1000,
+    });
+    expect(() => authService.verify(token)).toThrow(
+      'The token is valid, but the payload of the token is' +
+        ' not a TokenPayload object. This is ether a bug or a malicious attack.',
+    );
+  });
+  it('should throw Error("resourceOwnerId must be a number.")', () => {
+    const token = authService.sign({
+      userId: 0,
+      permissions: [],
+    });
+    expect(() => {
+      authService.audit(
+        token,
+        AuthorizedAction.other,
+        '1' as any as number,
+        'type',
+        1,
+      );
+    }).toThrow('resourceOwnerId must be a number.');
+  });
+  it('should throw Error("resourceId must be a number.")', () => {
+    const token = authService.sign({
+      userId: 0,
+      permissions: [],
+    });
+    expect(() => {
+      authService.audit(
+        token,
+        AuthorizedAction.other,
+        1,
+        'type',
+        '1' as any as number,
+      );
+    }).toThrow('resourceId must be a number.');
+  });
+  it('should throw AuthenticationRequiredError()', () => {
+    expect(() =>
+      authService.audit('', AuthorizedAction.other, 1, 'type', 1),
+    ).toThrow(new AuthenticationRequiredError());
+    expect(() =>
+      authService.audit(null!, AuthorizedAction.other, 1, 'type', 1),
```

Comment:
If the authService.audit is not exposed to user, the parameter should always be a string thanks to typescript's type check. So we should never worry about a null or undefined passing to it, unless the abuse of ! operator silenting the null check.

Reply:
NestJS framework will inject undefined if request does not contain something, even if you doesn't set it to optional.

For example, if you write
```typescript
@Headers('Authorization') auth: string
```
then, framework will inject an undefined into parameter auth, and it will be passed to verify(), and that is why verify() needs to do empty check.

Reply:
So the proper solution is to notate auth as `string | undefined` instead of passing `null!`

Reply:
this null! is used to simulate such situation. it's a test file, after all

Reply:
ok, ok, I fixed it https://github.com/SageSeekerSociety/cheese-backend/commit/b2f1f1f3851f81eb537879486fb97e4c81e76fec

Reply:
It's true that this is only a test. But if there is a case that undefined might be passed to this function, then the parameter type should be notated as optinoal.